pysat/_meta.py

```diff
@@ -609,100 +610,61 @@ def __eq__(self, other_meta):
-            # quick check on length
-            keys1 = [i for i in self.keys()]
-            keys2 = [i for i in other_meta.keys()]
-            if len(keys1) != len(keys2):
-                return False
-
-            # now iterate over each of the keys in the first one
-            # don't need to iterate over second one, if all of the first
-            # in the second we are good. No more or less items in second from
-            # check earlier.
-            for key in keys1:
-                if key not in keys2:
-                    return False
-
-            # do same checks on attributes
-            attrs1 = [i for i in self.attrs()]
-            attrs2 = [i for i in other_meta.attrs()]
-            if len(attrs1) != len(attrs2):
-                return False
-
-            for attr in attrs1:
-                if attr not in attrs2:
+            # Check if the variables and attributes are the same
+            for iter1, iter2 in [(self.keys(), other_meta.keys()),
+                                 (self.attrs(), other_meta.attrs())]:
+                list1 = [value for value in iter1]
+                list2 = [value for value in iter2]
+
+                try:
+                    testing.assert_lists_equal(list1, list2)
+                except AssertionError:
                     return False
 
-            # now check the values of all elements now that we know all
-            # variable and attribute names are the same
+            # Check that the values of all elements are the same. NaN is treated
+            # as equal, though mathematically NaN is not equal to anything
             for key in self.keys():
                 for attr in self.attrs():
-                    if not (self[key, attr] == other_meta[key, attr]):
-                        # np.nan is not equal to anything
-                        # if both values are NaN, ok in my book
-                        try:
-                            if not (np.isnan(self[key, attr])
-                                    and np.isnan(other_meta[key, attr])):
-                                # one or both are not NaN and they aren't equal
-                                # test failed
-                                return False
-                        except TypeError:
-                            # comparison above gets unhappy with string inputs
-                            return False
+                    if not testing.nan_equal(self[key, attr],
+                                             other_meta[key, attr]):
+                        return False
 
-            # check through higher order products
-            # in the same manner as code above
-            keys1 = [i for i in self.keys_nD()]
-            keys2 = [i for i in other_meta.keys_nD()]
-            if len(keys1) != len(keys2):
+            # Check the higher order products
+            keys1 = [key for key in self.keys_nD()]
+            keys2 = [key for key in other_meta.keys_nD()]
+            try:
```

Comment:
We should be able to do a call back into eq rather than check the higher order terms with duplicated code. It didn't work when I originally tried it but I was running up against a deadline. 
```
for key1, key2 in zip(keys1, keys2):
    if not key1 == key2:
        return False
```
and he lines below (until 664) can be dropped.

This isn't a have-to type comment, but an in case you are interested in reducing lines further type comment.

Reply:
Ok, I took a hack at this and couldn't get it to work.  I did learn something interesting about using NotImplemented in __eq__, though, so I made a small change.  Re-requesting review and living with the extra lines of code.